File: OLD Repos/Linear Algebra/Determinant/Algebraic Remainder of Sub-determinant.jl
#### Algebraic Remainder of Sub-determinant

# Input a n*n array as the determinant you want to analysis.
D = 
[
    4 1 5 4 3;
    3 4 6 2 5;
    2 7 5 1 3;
    1 6 4 3 6;
    2 1 6 5 4
]
n = 5

# The number of sub-determinant's ordinal:
k = 3
# Input the ordinals of rows you want to extract:
I = [1,3,5]
# Input the ordinals of columns you want to extract:
J = [2,3,4]

## Fortran-like code:
# Initialize the sub-determinant:
S = zeros(k,k)
# Initialize the remainder:
M = zeros(n-k,n-k)
# Initialize the algebraic remainder:
A = zeros(n-k,n-k)
# Assign values to the sub-determinant:
a = 1
b = 1
for i in I
    for j in J
        global S[a,b] = D[i,j]
        global b = b + 1
    end
    global b = 1
    global a = a + 1
end
println("The sub-determinant is ", S)

## Julia-like code:
# Assign values to the sub-determinant:
S = view(D,I,J)
println("The sub-determinant is ", S)

# Assign values to the remainder and algebraic remainder:
N = range(1, stop=n)
Ip = setdiff(N,I)
Jp = setdiff(N,J)
M = view(D,Ip,Jp)
println("The remainder is $M")
A = M*(-1)^(sum(I)+sum(J))
println("The algebraic remainder is $A")

File: Appendix Algebra/Determinant/Algebraic Remainder of Sub-determinant.py
#### Algebraic Remainder of Sub-determinant

#%%
import numpy as np

# Input a n*n array as the determinant you want to analysis.
D = np.array([  [4,1,5,4,3],
                [3,4,6,2,5],
                [2,7,5,1,3],
                [1,6,4,3,6],
                [2,1,6,5,4] ])
n = D.shape[0]

# The number of sub-determinant's ordinal:
k = 3
# Input the ordinals of rows you want to extract:
I = np.array([0,2,4])
# Input the ordinals of columns you want to extract:
J = np.array([1,2,3])

S = D[:,J][I,:]
S   # Sub-determinant
    # array([[1, 5, 4],
    #        [7, 5, 1],
    #        [1, 6, 5]])
N = np.arange(0,n,1)
Im = np.array(list(set(N).difference(set(I))))
Jm = np.array(list(set(N).difference(set(J))))
M = D[:,Jm][Im,:]
M   # The remainder
    # array([[3, 5],
    #        [1, 6]])
A = (-1)**(sum(I)+sum(J))*M
A   # The algebraic remainder
    # array([[3, 5],
    #        [1, 6]])


